parsers/KPX.py

```diff
@@ -196,115 +188,132 @@ def get_long_term_prod_data(
+        timed_data[date] = {"production": production_mix, "storage": storage_mix}
+
+    return timed_data
 
 
-def get_granular_real_time_prod_data(session: Optional[Session] = None) -> dict:
-    r0 = session or Session()
-    res_0 = r0.get(REAL_TIME_URL, verify=False)
-    chart_data = extract_chart_data(res_0.text)
+def get_granular_real_time_prod_data(
+    zone_key: ZoneKey = ZoneKey("KR"),
+    session: Session = Session(),
+    logger: Logger = getLogger(__name__),
+) -> List[dict]:
+    res = session.get(REAL_TIME_URL, verify=False)
+
+    production_list = ProductionBreakdownList(logger)
+
+    events = _parse_real_time_chart_data(res.text)
+    for dt, event in events.items():
+        production_list.append(
+            zoneKey=zone_key,
+            datetime=dt,
+            source=KR_SOURCE,
+            production=event["production"],
+            storage=event["storage"],
+        )
 
-    return chart_data
+    return production_list.to_list()
 
 
 @refetch_frequency(timedelta(minutes=5))
 def fetch_production(
-    zone_key: str = "KR",
-    session: Optional[Session] = None,
+    zone_key: ZoneKey = ZoneKey("KR"),
+    session: Session = Session(),
     target_datetime: Optional[datetime] = None,
     logger: Logger = getLogger(__name__),
 ) -> List[dict]:
-
-    if target_datetime is not None and target_datetime < arrow.get(
-        2021, 12, 22, 0, 0, 0, tzinfo=TIMEZONE
-    ):
+    first_available_date = arrow.get(2021, 12, 22, 0, 0, 0, tzinfo=TIMEZONE)
+    if target_datetime is not None and target_datetime < first_available_date:
         raise NotImplementedError(
             "This parser is not able to parse dates before 2021-12-22."
         )
 
-    if target_datetime is None:
-        target_datetime = arrow.now(TIMEZONE).datetime
-
-    all_data = []
+    _now = arrow.now(TIMEZONE)
 
-    if target_datetime.date() == arrow.now(TIMEZONE).date():
-        chart_data = get_granular_real_time_prod_data(session=session)
-
-        for datetime_key, chart_data_values in chart_data.items():
-            data = {
-                "zoneKey": "KR",
-                "datetime": datetime_key,
-                "capacity": {},
-                "production": {},
-                "storage": {},
-                "source": "https://new.kpx.or.kr",
-            }
-
-            data["storage"]["hydro"] = -chart_data_values["pumpedHydro"]
-
-            data["production"]["coal"] = chart_data_values["coal"]
-            data["production"]["gas"] = chart_data_values["gas"]
-            data["production"]["nuclear"] = chart_data_values["nuclear"]
-            data["production"]["oil"] = chart_data_values["oil"]
-            data["production"]["hydro"] = chart_data_values["hydro"]
-            data["production"]["unknown"] = chart_data_values["renewable"]
+    if target_datetime is None:
+        target_datetime = _now.datetime
 
-            all_data.append(data)
+    if target_datetime.date() == _now.date():
```

Comment:
I think it's a bit of a doggy comparison. If I understand correctly the thinking, if we don't pass a target datetime, we should use the real time source right ?
If that's the case you can eliminate this if by having  
```
if target_datetime is None:
  return get_real_time
 return get_granular_real_time_prod_data(
            zone_key=zone_key, session=session, logger=logger
        )

Reply:
Good catch - will simplify this